<launch>
    <!-- reactive_robot nodes to launch -->
    <node pkg="reactive_robot" type="autodrive_node" name="autodrive_node" output="screen"/>
    <node pkg="reactive_robot" type="collision_node" name="collision_node" output="screen"/>
    <node pkg="reactive_robot" type="obstacle_node" name="obstacle_node" output="screen"/>
    <node pkg="reactive_robot" type="decision_node" name="decision_node" output="screen"/>
    
    <!-- TurtleBot nodes to launch -->
    
    <arg name="world_file"  default="$(find reactive_robot)/worlds/hallway.world"/>

    <arg name="base"      value="$(optenv TURTLEBOT_BASE kobuki)"/> <!-- create, roomba -->
    <arg name="battery"   value="$(optenv TURTLEBOT_BATTERY /proc/acpi/battery/BAT0)"/>  <!-- /proc/acpi/battery/BAT0 --> 
    <arg name="gui" default="true"/>
    <arg name="stacks"    value="$(optenv TURTLEBOT_STACKS hexagons)"/>  <!-- circles, hexagons --> 
    <arg name="3d_sensor" value="$(optenv TURTLEBOT_3D_SENSOR kinect)"/>  <!-- kinect, asus_xtion_pro --> 

    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="use_sim_time" value="true"/>
        <arg name="debug" value="false"/>
        <arg name="gui" value="$(arg gui)" />
        <arg name="world_name" value="$(arg world_file)"/>
    </include>
  
    <include file="$(find turtlebot_gazebo)/launch/includes/$(arg base).launch.xml">
        <arg name="base" value="$(arg base)"/>
        <arg name="stacks" value="$(arg stacks)"/>
        <arg name="3d_sensor" value="$(arg 3d_sensor)"/>
    </include>
  
    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
        <param name="publish_frequency" type="double" value="30.0" />
    </node>
  
    <!-- Fake laser -->
    <node pkg="nodelet" type="nodelet" name="laserscan_nodelet_manager" args="manager"/>
    <node pkg="nodelet" type="nodelet" name="depthimage_to_laserscan"
            args="load depthimage_to_laserscan/DepthImageToLaserScanNodelet laserscan_nodelet_manager">
        <param name="scan_height" value="10"/>
        <param name="output_frame_id" value="/camera_depth_frame"/>
        <param name="range_min" value="0.45"/>
        <remap from="image" to="/camera/depth/image_raw"/>
        <remap from="scan" to="/scan"/>
    </node>

    <!-- Keyboard control -->
    <remap from="cmd_vel_mux/input/teleop" to="reactive_robot/keyboard_input"/>
    <include file="$(find turtlebot_teleop)/launch/keyboard_teleop.launch"/>


    <!-- Disable console prints -->
    <remap from="rosout" to="rosout_throwaway"/>

    <!-- SLAM mapping -->
    <include file="$(find turtlebot_navigation)/launch/includes/gmapping/$(arg 3d_sensor)_gmapping.launch.xml"/>

</launch>